### ISSUE TITLE
Prevent storagenode from starting in empty/invalid storage directory, unless explicitly stated.

### ISSUE THREAD

**Issue by nizedk** [storj/storj#3973]
Hi team,

Feature request, based on numerous stories of DQ'ed nodes based on misconfigurations;

The storagenode shouldn't start if the data directory isn't populated as expected. Typically by moving nodes/disks/filesystem, the SNO, or a missing mount option in fstab, or other accidental omission altered the paths/their contents. The node should check, state it at startup, and shut down, before being reachable from uplinks and sattelites.

What happens now if that happens; the node quickly fails audit checks, and gets DQ'ed silently, unrelated to typing error, missing mountpoint, hardware error or other human mistake. Silent DQ after the operator saw the node start, expecting things to be in order.

When initializing a new node, an option should be set once, or another set of checks should happen, to certify that yes, this node is new and empty. Maybe clicking something in the web dashboard, like writing 'yes this is a new node without data' or something, or easier; in the CLI/docker options,  unless a '--first-time' or '--new-node-empty-storage' has been specified, simply omit the directory check, and proceeds as has happened until now.

The whole DQ level and being suspended treshold suggestion is great and valid, but only AFTER the most common reason for DQ is removed; SNO operators inadvertently starting their node with an incorrect data directory.

The check could also to verify that it is a previously unused storage node identity, if the data directory is empty. If the ID has been seen before, shutdown if data directory isn't containing something relevant to the identity.

Maybe write a file containing the identity in the root, so that the relative positioning in a filesystem is verified pr data/logs/identiy or any other important path is also verified by this. Simple 'if exists' checks would be better than now.

Keep doing great work on the project, you are behind an awesome project. Much appreciated.

**Comment by Pentium100MHz** [comment]
I would like to add a suggestion:
Upon startup, the node should calculate the space used by the data files (`du ...`) and compare the result to the number stored in the database. If the files take up significantly less space than they should, maybe the transfer was incomplete or something and the node should crash. This should be possible to override with a command line argument.

Also the node should crash if it was started with the "new node" argument and the data directory is not empty.

**Comment by nizedk** [comment]
This initial feature request was specifically to prevent the simplest of errors; the 'wrong path' type, based on omission or error, or relative positioning; ie. /var/storj/storj vs /var/storj/  etc.

On some nodes it would take a very long time to enumerate all blobs with 'du'; that's the reasoning behind the suggestion for a simple 'placeholder' file that should be present one or more expected places.

It might be a good idea to check for 'one of the new' and 'one of the old' blobs, if the DB allows for such a query. Maybe pick out 100 files at random from the DB, verify they are present in the filesystem, and continue booting the node if they are, as opposed to traversing potentially 10s of millions of files?

Thank you for input, there are many ways to improve the nodes. :-)

**Comment by pacproduct** [comment]
I think it'd be nice to have this `--init` option, but the way I see it it should start the node in a special mode where it would initialize stuff, prepare databases, identifiers and whatnot, and then **stop** with a success message (if the target directory was adapted, accessible, not occupied by an existing node and so forth...).
Just my 2 cts :)

**Comment by littleskunk** [comment]
Take a look at this commit: [+ 5a337c48] {cmd,private,storagenode}: create storage dir verification during setup

It will be included in the next release.

**Comment by pacproduct** [comment]
@littleskunk What triggers the "setup" procedure? I mean, how does the Node know it shouldn't start its "setup" procedure when targeting an empty directory?

**Comment by littleskunk** [comment]
@pacproduct the operator has to call setup. See documentation.

Issues was fixed with the commit I mentioned above. -> Closed.
